### ISSUE TITLE
Task dependencies / before/after hooks

### ISSUE THREAD

**Issue by bitprophet** [fabric/fabric#452]
1.3 features `execute()` which allows for meta-tasks which call other tasks and honors their individual `@hosts`, etc settings. This solves many of the more complicated multi-task invocation problems people have run into.

However, there are some arguments in favor of having a flipped-around version of a meta-task, namely specifying that Task A should always execute Task B before or after it runs. (Versus creating a Task C consisting solely of `execute(Task B); execute(Task A)`.) This is commonly referred to as "dependencies" or a "call chain", as seen in Rake/Capistrano/etc.

One such argument is just that "good" refactored design of tasks means you may often want to call 'subtasks' stand-alone, so "just put your pre-requisite in a meta-task" doesn't help. E.g. setting up an "environment" (in the staging/production sense) before executing.

I was -0 on this before because *technically* you can solve this solely with `execute`, but in more complex cases that starts to require a "runner" task that takes the real task to run as an argument. At that point, decorating a number of subtasks with "requires X to be run beforehand" is more Pythonic.

----

Closely related: the idea of having a queue instead of a loop in `fabric.main.main`, as mentioned in #391. In such a system, dependency decorators might act by simply loading the referenced items into the queue before/after the decorated task.

**Comment by bitprophet** [comment]
This is implemented in Invoke already (happened last year-ish) so it'll definitely be available for Fabric 2.0 users.

**Comment by wardi** [comment]
@bitprophet should this be working on the v2 branch right now? I can use `@task(pre=...)` with invoke but it doesn't seem to work when I run from the fabric script. Should I be using fabric only as a library?

**Comment by bitprophet** [comment]
@wardi Think that's a bug/TODO, just confirmed it on my end, suspect it has to do with Fabric 2 having a custom executor subclass. Should be relatively easily fixed, will see if I can bang it out real quick.

**Comment by bitprophet** [comment]
Ah yea - seeing the code in question (& its comments) reminds me why this was deferred, it's a more complex question in Fabric than in regular Invoke, because of host parameterization.

Ye olde "task `foo` depends on task `bar` is, in Invoke, pretty easy: just call `bar, then `foo`.

In Fabric, if `foo` is being run with a host list of `A,B,C`, do we run 'bar, foo-A, foo-B, foo-C' _or_ do we run 'bar-A, foo-A, bar-B, foo-B, bar-C, foo-C'? (And the right answer there will differ between use cases as well.)

So right now, Fabric 2 decided to not even try answering that question and **ignores** pre/post.

There's a bunch of existing commentary about this in pyinvoke/invoke#461 and I even have a branch open with unfinished work towards a call-graph, pure-dependency-based setup. But barring completion of that, I'll have to decide on the most-applicable interpretation of the above problem as it applies to `pre`/`post`, and apply that to Fabric 2's executor class so it works for at least some use cases instead of none 🙃

**Comment by bitprophet** [comment]
FWIW, if I had to just "do it well enough" today, I'd go for the "pre-tasks once, main task N times across hosts, post-tasks once" version of the scenario, since the other scenario can be effected well-enough by users calling their pre/post tasks manually within the main task's body.

Of note, this also brings pyinvoke/invoke#261 into play (especially as pertaining to the hosts-list factor, since it's not even technically part of a task's signature but is a sort of metadata.) Complicated!

**Comment by wardi** [comment]
@bitprophet thanks! Ignoring pre/post seems to make sense. Refuse to guess and all that.

I've gone the simple route and just call the pre function and implement my own check based on a context variable to see if it has already been run in the same session.

**Comment by bitprophet** [comment]
@wardi I've actually just pushed a change to the v2 branch that implements what I said I would do, above, re: `pre -> main-host-1 -> ... -> main-host-N -> post`. Since I would've had to do it anyways if I wanted to defer the "real" fix, and I had the patient on the table from simply looking at the problem & writing tests, I figured why not spend the extra 5 minutes.

Let me know how that works for you, if you've time to pull it down and check it out!

**Comment by wardi** [comment]
not exactly, I was hoping that I could use

```python
@task
def this_first(c): ...

@task(this_first)
def thing1(c): ...

@task(this_first)
def thing2(c): ...
```

to do
```
$ fab thing1 thing2
this_first
thing1
thing2
```

My specific "this_first" uses are "have the user enter their ssh password to set up c.config" and "pull all local git repos before they are deployed to various targets (in procedures given by thing1 and thing2)"

But as I said, I'm fine with implementing this in plain python instead of using pre/post if this isn't a common use case.

**Comment by bitprophet** [comment]
Ah yea, that's part of the other outstanding issues re: real dependency graph tracking. The current setup is old and very literal/naive. So, keep an eye on pyinvoke/invoke#461!
